src/main.rs

```diff
@@ -1,3 +1,192 @@
-fn main() {
-    println!("Hello, world!");
+extern crate futures;
+extern crate reqwest;
+extern crate tokio;
+
+use futures::Future;
+use reqwest::r#async::Client;
+
+trait InfluxDbQuery {
+    fn build<'a>(self) -> String;
 }
+
+impl InfluxDbQuery {
+    pub fn write() -> InfluxDbWrite {
+        InfluxDbWrite {
+            _measurement: String::from("marina_3"),
+            _fields: Vec::new(),
+            _tags: Vec::new(),
+        }
+    }
+
+    // pub fn read() {}
+}
+
+pub struct InfluxDbWrite {
+    _fields: Vec<(String, String)>,
+    _tags: Vec<(String, String)>,
+    _measurement: String,
+    // precision: Precision
+}
+
+impl InfluxDbWrite {
+    fn add_field<'a, S>(&'a mut self, point: S, value: S) -> &'a mut Self
+    where
+        S: Into<String>,
+    {
+        self._fields.push((point.into(), value.into()));
+        self
+    }
+
+    fn add_tag<'a, S>(&'a mut self, tag: S, value: S) -> &'a mut Self
+    where
+        S: Into<String>,
+    {
+        self._tags.push((tag.into(), value.into()));
+        self
+    }
+}
+
+impl InfluxDbQuery for InfluxDbWrite {
+    // fixme: time (with precision) and measurement
+    fn build<'a>(self) -> String {
+        let tags = self
+            ._tags
+            .into_iter()
+            .map(|(tag, value)| format!("{tag}={value}", tag = tag, value = value))
+            .collect::<Vec<String>>()
+            .join(",");
+        let fields = self
+            ._fields
+            .into_iter()
+            .map(|(field, value)| format!("{field}={value}", field = field, value = value))
+            .collect::<Vec<String>>()
+            .join(",");
+
+        format!(
+            "measurement,{tags} {fields} time",
+            tags = tags,
+            fields = fields
+        )
+    }
+}
+
+pub struct InfluxDbClient {
+    _url: String,
+    _database: String,
+    // _auth: InfluxDbAuthentication | NoAuthentication
+}
+
+pub fn main() {}
+
+impl InfluxDbClient {
+    pub fn ping(&self) -> impl Future<Item = (String, String), Error = ()> {
+        Client::new()
```

Comment:
Why are you creating a new `Client` everytime?

Reply:
How would you solve this?

Reply:
Well, generally I prefer to have the `Client` be stored in a variable of the associated type and give the user a way to specify their own `Client` if they prefer. This might also be helpful when trying to mock http calls